### PR TITLE
Building singlefilehost with clr partition.

### PR DESCRIPTION
**Pull request by VSadov**
- single file host builds with clr partition
- removed hacks that align single file host build flavor and CRT to match with clr
- host build script now just picks singlefilehost executable from clr artifacts
- fixed a few places which introduce duplicate definitions when JIT and VM are linked together. 

NOTE: static libs for the native libraries that are built in clr partitions are still installed into artifacts. 
I left them be, in case there are other uses, but singlefilehost no longer needs them.

Resolves:https://github.com/dotnet/runtime/issues/43700

### CODE REVIEW

File: src/coreclr/CMakeLists.txt [hunk 1]
```diff
@@ -126,6 +126,10 @@ if(FEATURE_STANDALONE_GC)
   endif (CLR_CMAKE_BUILD_SUBSET_RUNTIME)
 endif(FEATURE_STANDALONE_GC)
 
+if(NOT CLR_CROSS_COMPONENTS_BUILD)
+  add_subdirectory(${CLR_REPO_ROOT_DIR}/src/installer/corehost/cli/apphost/static Corehost.Static)
```
Comment (am11) [inline src/coreclr/CMakeLists.txt:130]: nit: after the rebase, this can be simplified to:
```suggestion
  add_subdirectory(${CLR_ENG_NATIVE_DIR}/corehost/apphost/static Corehost.Static)
```

File: src/coreclr/build-runtime.sh [hunk 2]
```diff
@@ -250,6 +280,14 @@ __CMakeArgs="-DCLR_CMAKE_PGO_INSTRUMENT=$__PgoInstrument -DCLR_CMAKE_OPTDATA_PAT
 __CMakeArgs="-DCLR_CMAKE_BUILD_SUBSET_JIT=$__BuildJit -DCLR_CMAKE_BUILD_SUBSET_ALLJITS=$__BuildAllJits -DCLR_CMAKE_BUILD_SUBSET_RUNTIME=$__BuildRuntime $__CMakeArgs"
 __CMakeArgs="-DCLR_CMAKE_BUILD_TESTS=$__BuildPALTests $__CMakeArgs"
 
+__CMakeArgs="-DCLI_CMAKE_HOST_VER=$__host_ver -DCLI_CMAKE_COMMON_HOST_VER=$__apphost_ver -DCLI_CMAKE_HOST_FXR_VER=$__fxr_ver $__CMakeArgs"
+__CMakeArgs="-DCLI_CMAKE_HOST_POLICY_VER=$__policy_ver -DCLI_CMAKE_PKG_RID=$__DistroRid -DCLI_CMAKE_COMMIT_HASH=$__commit_hash $__CMakeArgs"
+__CMakeArgs="-DFEATURE_DISTRO_AGNOSTIC_SSL=$__PortableBuild $__CMakeArgs"
+
+if [[ "$__PortableBuild" == 1 ]]; then
+    __CMakeArgs="-DCLI_CMAKE_PORTABLE_BUILD=1 $__CMakeArgs"
```
Comment (am11) [inline src/coreclr/build-runtime.sh:288]: nit: `CLI_CMAKE_PORTABLE_BUILD` is used in one place to define `FEATURE_PORTABLE_BUILD`:
https://github.com/dotnet/runtime/blob/ddf94a71c7606aa433f4776841afd354a33b0502/src/native/corehost/setup.cmake#L4-L6
but according to `git log -SFEATURE_PORTABLE_BUILD` results, `FEATURE_PORTABLE_BUILD` was never used (since it was introduced https://github.com/dotnet/runtime/commit/4f68af44197d0631e3f45673873dd6a69798a1e8). We can delete them both.

Comment (VSadov) [reply src/coreclr/build-runtime.sh:288]: I think `__PortableBuild` is entirely uninteresting here. 
It basically controls how SSL builds in libraries by setting `FEATURE_DISTRO_AGNOSTIC_SSL`, but that is also done in `build-commons.sh`

Comment (am11) [reply src/coreclr/build-runtime.sh]: Yup. It is unused in corehost's Unix shell scripts (because distro specific RID for source-build is taken care by `build-commons.sh`), and for Windows, I think for a while now, nobody is building or testing non-portable flavors.  cm_BaseRid is always overwritten on line 151 https://github.com/dotnet/runtime/blob/39afd1f46f8f03582b0414591ff7a1e2f1809979/src/native/corehost/build.cmd#L146-L151 the default is `true` https://github.com/dotnet/runtime/blob/39afd1f46f8f03582b0414591ff7a1e2f1809979/Directory.Build.props#L95-L96 and CI's YAML configs also set it to true (to make sure it is really ture 🙂).

**Comment by VSadov** [comment]
I think this is ready for review.

File: src/coreclr/jit/CMakeLists.txt [hunk 3]
```diff
@@ -348,14 +348,14 @@ else()
   clr_unknown_arch()
 endif()
 
-set(SOURCES
+set(JIT_CORE_SOURCES
```
Comment (VSadov) [inline src/coreclr/jit/CMakeLists.txt:351]: This change was needed to not include resources in the statically linked JIT.

File: src/native/corehost/apphost/static/CMakeLists.txt [hunk 4]
```diff
@@ -42,6 +42,20 @@ include(../../fxr/files.cmake)
 include(../../hostpolicy/files.cmake)
 include(../../hostcommon/files.cmake)
 
+if(MSVC)
+    add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/wd4996>)
```
Comment (VSadov) [inline src/native/corehost/apphost/static/CMakeLists.txt:46]: these `wd` defined when everything in host.native is built. 
Since we build this separately from others now, I had to add the options here.

File: src/native/corehost/apphost/static/CMakeLists.txt [hunk 5]
```diff
@@ -84,59 +98,18 @@ if(CLR_CMAKE_TARGET_WIN32)
     # Disable manifest generation into the file .exe on Windows
     add_linker_flag("/MANIFEST:NO")
 
-    get_property(ALL_COMPILE_OPTIONS TARGET ${PROJECT_NAME} PROPERTY COMPILE_OPTIONS)
-    string(TOUPPER ${RUNTIME_CONFIG} UPPERCASE_RUNTIME_CONFIG)
-
-    # make sure that certain compiler and linker settings match the runtime config
```
Comment (VSadov) [inline src/native/corehost/apphost/static/CMakeLists.txt:90]: I am very happy that this is gone. It seemed very fragile.

File: src/native/corehost/hostmisc/pal.windows.cpp [hunk 6]
```diff
@@ -423,7 +423,7 @@ bool pal::get_global_dotnet_dirs(std::vector<pal::string_t>* dirs)
 
 // To determine the OS version, we are going to use RtlGetVersion API
 // since GetVersion call can be shimmed on Win8.1+.
-typedef NTSTATUS (WINAPI *pFuncRtlGetVersion)(RTL_OSVERSIONINFOW *);
+typedef LONG (WINAPI *pFuncRtlGetVersion)(RTL_OSVERSIONINFOW *);
```
Comment (VSadov) [inline src/native/corehost/hostmisc/pal.windows.cpp:426]: I did not want to include `windows.h` for this. Host include that.

**Comment by VSadov** [comment]
I do not think `runtime-staging` is failing because of this change. It looks like it is failing in other PRs too.

File: src/coreclr/dlls/mscoree/coreclr/CMakeLists.txt [hunk 7]
```diff
@@ -1,3 +1,10 @@
+
+# we need to define this library before adding .def fiels to sources.
```
Comment (janvorli) [inline src/coreclr/dlls/mscoree/coreclr/CMakeLists.txt:2]: I wonder why it wasn't needed before.

File: src/native/corehost/apphost/apphost.windows.cpp [hunk 8]
```diff
@@ -6,6 +6,7 @@
 #include "pal.h"
 #include "trace.h"
 #include "utils.h"
+#include <shellapi.h>
```
Comment (janvorli) [inline src/native/corehost/apphost/apphost.windows.cpp:9]: Is this needed?

File: src/native/corehost/build.sh [hunk 9]
```diff
@@ -117,3 +107,8 @@ check_prereqs
 
 # Build the installer native components.
 build_native "$__TargetOS" "$__BuildArch" "$__scriptpath" "$__IntermediatesDir" "$__CMakeArgs" "installer component"
+
+if [[ "$__RuntimeFlavor" != "Mono" ]]; then
+    echo Copying "$__CoreClrArtifacts/corehost/."  to "$__CMakeBinDir/corehost"
+    cp -a "$__CoreClrArtifacts/corehost/."  "$__CMakeBinDir/corehost"
```
Comment (janvorli) [inline src/native/corehost/build.sh:113]: Why do we  copy all files from corehost on Unix while on Windows, we copy the .exe and .pdb explicitly?

File: src/native/corehost/hostmisc/pal.windows.cpp [hunk 10]
```diff
@@ -434,7 +434,7 @@ pal::string_t pal::get_current_os_rid_platform()
     // Init the buffer
     ZeroMemory(&osinfo, sizeof(osinfo));
     osinfo.dwOSVersionInfoSize = sizeof(osinfo);
-    HMODULE hmodNtdll = LoadLibrary("ntdll.dll");
+    HMODULE hmodNtdll = LoadLibraryA("ntdll.dll");
```
Comment (janvorli) [inline src/native/corehost/hostmisc/pal.windows.cpp:437]: Was this change needed?

File: src/native/corehost/hostmisc/pal.windows.cpp (hunk 6 above)
Comment (janvorli) [reply src/native/corehost/hostmisc/pal.windows.cpp:426]: I am not sure I understand this - the pal.windows.cpp set of includes didn't seem to have changed.

File: src/native/corehost/setup.cmake [hunk 11]
```diff
@@ -1,44 +1,52 @@
...
+    # CLR partition builds only the single file host where hosting components are all statically linked.
+    # the versioning information is irrelevant and may only come up in tracing.
+    # so we will use "static"
+    add_definitions(-DHOST_POLICY_PKG_VER="static")
+    add_definitions(-DHOST_FXR_PKG_VER="static")
+    add_definitions(-DHOST_PKG_VER="static")
+    add_definitions(-DCOMMON_HOST_PKG_VER="static")
+    add_definitions(-DHOST_POLICY_PKG_NAME="static")
+    add_definitions(-DHOST_POLICY_PKG_REL_DIR="static")
+    add_definitions(-DREPO_COMMIT_HASH="static")
```
Comment (janvorli) [inline src/native/corehost/setup.cmake:14]: I wonder if this change won't add "static" as another commit hash to the host (besides the one coming from coreclr). What does `strings singlefilehost | grep "@(#)"` show? It should print just one string with the real commit hash.

File: src/native/corehost/hostmisc/pal.windows.cpp (hunk 10 above)
Comment (VSadov) [reply src/native/corehost/hostmisc/pal.windows.cpp:437]: coreclr defines LoadLibrary as LoadLibraryW. Without the change I would have to change the string literal to `_X("ntdll.dll")`, but then the standalone hosts would not build because they map LoadLibrary to LoadLibraryA.

Since we use a char string here, I assumed we expect LoadLibraryA to be called, so I made it more explicit.

---
In reply to: [578794607](https://github.com/dotnet/runtime/pull/48254#discussion_r578794607) [](ancestors = 578794607)

File: src/native/corehost/hostmisc/pal.windows.cpp (hunk 6 above)
Comment (VSadov) [reply src/native/corehost/hostmisc/pal.windows.cpp:426]: I think this is because corehost.proj adds `#include windows.h`  to the native versions file and now we do not include the file.

---
In reply to: [578795224](https://github.com/dotnet/runtime/pull/48254#discussion_r578795224) [](ancestors = 578795224)

Comment (VSadov) [reply src/native/corehost/hostmisc/pal.windows.cpp:426]: Actually we never included that file in singlefilehost, since it is built with `set(SKIP_VERSIONING 1)`, so now I am not sure.
Somehow NTSTATUS is defined when building installer partition and not when building coreclr. 

And that is probably the same reason why `#include <shellapi.h>` is needed to be included in `pal.windows.cpp`. 

---
In reply to: [578819238](https://github.com/dotnet/runtime/pull/48254#discussion_r578819238) [](ancestors = 578819238,578795224)

File: src/native/corehost/setup.cmake (hunk 11 above)
Comment (VSadov) [reply src/native/corehost/setup.cmake:14]: This is just a string that could be seen in tracing. 
Standalone host also has `hostfxr_get_dotnet_environment_info` API, but singlefile host does not expose hosting APIs. The final stitched together executable is not supposed to act as a host - it is just a user program. 

---
In reply to: [578796864](https://github.com/dotnet/runtime/pull/48254#discussion_r578796864) [](ancestors = 578796864)

Comment (VSadov) [reply src/native/corehost/setup.cmake:14]: The format of this string literal is like "a874f3adcb3f159edb077ea9bb7d841391ed7465"  - no `#`

---
In reply to: [578823880](https://github.com/dotnet/runtime/pull/48254#discussion_r578823880) [](ancestors = 578823880,578796864)

Comment (janvorli) [reply src/native/corehost/setup.cmake:14]: Got it, I've thought it is used by the version string stored in our executables.

File: src/coreclr/dlls/mscoree/coreclr/CMakeLists.txt (hunk 7 above)
Comment (VSadov) [reply src/coreclr/dlls/mscoree/coreclr/CMakeLists.txt:2]: This is ignored by a static lib, but becomes an error if it is `coreclr_static` is an object lib. 
Same about duplicate inclusion of `gcdecode.cpp`, which causes a warning with static lib and an error with object lib. 

Basically static lib can tolerate these and we do not _have_ to fix. 
I had to do the fixes for the object lib, but even though we are not using it in this change, I kept the fixes, since these things (like including .def files) do not look intentional.

File: src/native/corehost/apphost/apphost.windows.cpp (hunk 8 above)
Comment (VSadov) [reply src/native/corehost/apphost/apphost.windows.cpp:9]: It is needed for the 
```c
            // Open the URL in default browser
            ::ShellExecuteW(
```

The question is - why it was not needed before...

File: src/native/corehost/build.sh (hunk 9 above)
Comment (VSadov) [reply src/native/corehost/build.sh:113]: We need to copy the entire "/corehost". There are only two files there on Windows - `singlefilehost.exe` and `PDB/singlefilehost.pdb`

Unix has a convenient command for copying the content of the entire `/corehost` and cmd alternative `xcopy` is harder to use - it gets confused whether we are copying a file or a folder, whether the destination has to exist or must not exist, and so on.

`copy` with individual 2 files ended up a lot more reliable.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

File: src/native/corehost/apphost/apphost.windows.cpp (hunk 8 above)
Comment (VSadov) [reply src/native/corehost/apphost/apphost.windows.cpp:9]: On why this include was not needed when this was building in installer partition. - clr partition build defines `WIN32_LEAN_AND_MEAN` and installer partition does not.

**Comment by VSadov** [comment]
Thanks!!!

**Comment by BruceForstall** [comment]
@VSadov It appears this change broke the JIT rolling build (which only builds the JIT, using, for example, `src/coreclr/build-runtime.cmd checked x64 -ci -skipruntime -nopgooptimize -skiprestoreoptdata`):

https://dev.azure.com/dnceng/internal/_build/results?buildId=1004132&view=results

cc @dotnet/jit-contrib

**Comment by am11** [comment]
@BruceForstall, I have opened #48552 with the fix. The condition just need to convey that the building subset is CLR.
